Physical_Engine/Create/ShapeCode/BS8666/ShapeCode14.cs

```diff
@@ -0,0 +1,53 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2021, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.ComponentModel;
+using BH.oM.Geometry;
+using BH.oM.Physical.Reinforcement.BS8666;
+using BH.oM.Reflection.Attributes;
+
+namespace BH.Engine.Physical
+{
+    public static partial class Create
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Creates a ShapeCode object using the parameters provided. Refer to the object descriptions for alignment.")]
+        [Output("shapeCode", "A ShapeCode to be used with Reinforcement objects.")]
+        public static ShapeCode14 ShapeCode14(double a, double b, double c, double d)
```

Comment:
This create method is allowing you to create an invalid object (where a^2 != b^2+d^2).

To me at least, the sole reason of making the shape codes immutable is to avoid this.

Would not keep this create method, but instead have some methods that calculates what needs to be calculated. For example (not sure if these are the exactly right ones):

- take b,c,d -> calculate a
- take a,c and angle -> calculate a and d
- take a,b,c -> calculate d

not sure we need all of them, but just to exemplify.

When thinking of it, I guess you kind of need the diameter and bend radius to do this. Wonder then if it is worth moving those properties to the shape code objects as well.

That would also actually help simplify some bits of your code. For example the `IsValid` and `Centreline` methods that currently requires the shape code object as well as those two parameter to be able to check the object/generate the curve. The more I think of it, I think it is worth promoting up those properties just for that reason.